app/src/test/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmServicesTest.kt

```diff
@@ -0,0 +1,97 @@
+    }
+
+    @Test
+    fun `scheduleEventAlarm only invokes "set" when "discardExisting" is "false"`() {
+        val onPendingIntentBroadcast: PendingIntentCallback = { context, requestCode, intent, flags ->
+            assertThat(context).isEqualTo(mockContext)
+            assertThat(requestCode).isEqualTo(alarm.eventId.toInt())
+            assertIntentExtras(intent, AlarmReceiver.ALARM_LECTURE)
+            assertThat(flags).isEqualTo(0)
+            pendingIntent
+        }
+        val alarmServices = AlarmServices(alarmManager, onPendingIntentBroadcast)
+        alarmServices.scheduleEventAlarm(mockContext, alarm, false)
+        verify(alarmManager, never()).cancel(pendingIntent)
+        verify(alarmManager, once()).set(AlarmManager.RTC_WAKEUP, alarm.startTime, pendingIntent)
+    }
+
+    @Test
+    fun `discardEventAlarm invokes "cancel"`() {
+        val onPendingIntentBroadcast: PendingIntentCallback = { context, requestCode, intent, flags ->
+            assertThat(context).isEqualTo(mockContext)
+            assertThat(requestCode).isEqualTo(alarm.eventId.toInt())
+            assertIntentExtras(intent, AlarmReceiver.ALARM_DELETE)
+            assertThat(flags).isEqualTo(0)
+            pendingIntent
+        }
+        val alarmServices = AlarmServices(alarmManager, onPendingIntentBroadcast)
+        alarmServices.discardEventAlarm(mockContext, alarm)
+        verify(alarmManager, once()).cancel(pendingIntent)
+    }
+
+
+    @Test
+    fun `discardAutoUpdateAlarm invokes "cancel"`() {
+        val onPendingIntentBroadcast: PendingIntentCallback = { context, requestCode, intent, flags ->
+            assertThat(context).isEqualTo(mockContext)
+            assertThat(requestCode).isEqualTo(0)
+            assertThat(intent.component!!.className).isEqualTo(AlarmReceiver::class.java.name)
+            assertThat(intent.action).isEqualTo(AlarmReceiver.ALARM_UPDATE)
+            assertThat(flags).isEqualTo(0)
+            pendingIntent
+        }
+        val alarmServices = AlarmServices(alarmManager, onPendingIntentBroadcast)
+        alarmServices.discardAutoUpdateAlarm(mockContext)
+        verify(alarmManager, once()).cancel(pendingIntent)
+    }
+
+    // TODO Move into a unit test for AlarmReceiver once it is written.
+    private fun assertIntentExtras(intent: Intent, action: String) {
+        assertThat(intent.getIntExtra(BundleKeys.ALARM_DAY, 9)).isEqualTo(alarm.day)
+        assertThat(intent.getStringExtra(BundleKeys.ALARM_LECTURE_ID)).isEqualTo(alarm.eventId)
+        assertThat(intent.getLongExtra(BundleKeys.ALARM_START_TIME, 0)).isEqualTo(alarm.startTime)
+        assertThat(intent.getStringExtra(BundleKeys.ALARM_TITLE)).isEqualTo(alarm.eventTitle)
+        assertThat(intent.component!!.className).isEqualTo(AlarmReceiver::class.java.name)
+        assertThat(intent.action).isEqualTo(action)
+        assertThat(intent.data).isEqualTo(Uri.parse("alarm://${alarm.eventId}"))
+    }
+
+    private fun once() = times(1)
```

Comment:
move member to the top please

Reply:
It is a helper function, not a member. Or am I missing something here?

Reply:
you are right. but its defined after usages, this is what I was concerned about :)

Reply:
No problem. The JVM handles that easily. I try to structure code in a way that unimportant functions are at the bottom so they do not steal the focus of the reader.